### PR TITLE
Allow GCS to download single file

### LINKED ISSUE #4013: Storage Initializer GCS no model found with single file
/kind bug

**What steps did you take and what happened:**
When GCS storage url points to single file directly (eg. `gs://REDACTED/model/test-folder/model.pb`), following error occurs:

```
Traceback (most recent call last):
  File "/storage-initializer/scripts/initializer-entrypoint", line 17, in <module>
    Storage.download(src_uri, dest_path)
  File "/kserve/kserve/storage/storage.py", line 89, in download
    model_dir = Storage._download_gcs(uri, out_dir)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/kserve/kserve/storage/storage.py", line 353, in _download_gcs
    raise RuntimeError("Failed to fetch model. No model found in %s." % uri)
RuntimeError: Failed to fetch model. No model found in gs://REDACTED/model/test-folder/model.pb.
```

**What did you expect to happen:**
It should download the single file directly

**What's the InferenceService yaml:**
```
apiVersion: "serving.kserve.io/v1beta1"
kind: "InferenceService"
metadata:
  name: "flower-gcs"
  annotations:
    serving.kserve.io/enable-metric-aggregation:  "true"
    serving.kserve.io/enable-prometheus-scraping: "true"
spec:
  predictor:
    tensorflow:
      storageUri: "gs://REDACTED/model/test-folder/model.pb"
```

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Istio Version:
- Knative Version:
- KServe Version:
- Kubeflow version:
- Cloud Environment:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube/Kind version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by spolti**
fixes 	https://github.com/kserve/kserve/issues/4013

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note

```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

### CODE REVIEW

**Comment by lizzzcai** [comment]
@bentohset can you check if this can fix your issue.

**Comment by bentohset** [comment]
@spolti I was in the midst of making a PR for the issue but thanks for working on this! Took a look at your changes, tested them and is working as intended. However, the unit tests are not passing. Also could you add a unit test case for single file downloads? Thanks!

File: python/storage-initializer/scripts/model.joblib
Comment (spolti) [inline python/storage-initializer/scripts/model.joblib:1]: @andyi2it this file as added as part of [Implement Huggingface model download in storage initializer (](https://github.com/kserve/kserve/commit/1e28d4ffb9294fd5fd1c377d57cb9ad2a3be17d7)

Is it needed or can I delete it?

**Comment by spolti** [comment]
cc @yuzisun @terrytangyuan

**Comment by bentohset** [comment]
@spolti Did another test of your changes and seem like the following inferenceservice produces an error:
```yaml
apiVersion: "serving.kserve.io/v1beta1"
kind: "InferenceService"
metadata:
  name: "flower-gcs"
  annotations:
    serving.kserve.io/enable-metric-aggregation:  "true"
    serving.kserve.io/enable-prometheus-scraping: "true"
spec:
  predictor:
    tensorflow:
      storageUri: "gs://REDACTED/model/test-folder/model.pb"
```

error:
```bash
2024-10-28 06:46:56.030 1 kserve INFO [initializer-entrypoint:<module>():16] Initializing, args: src_uri [gs://REDACTED/model/test-folder/model.pb] dest_path[ [/mnt/models]
2024-10-28 06:46:56.030 1 kserve INFO [storage.py:download():64] Copying contents of gs://REDACTED/model/test-folder/model.pb to local
Traceback (most recent call last):
  File "/storage-initializer/scripts/initializer-entrypoint", line 17, in <module>
    Storage.download(src_uri, dest_path)
  File "/kserve/kserve/storage/storage.py", line 89, in download
    model_dir = Storage._download_gcs(uri, out_dir)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/kserve/kserve/storage/storage.py", line 372, in _download_gcs
    raise RuntimeError("Failed to fetch model. No model found in %s." % uri)
RuntimeError: Failed to fetch model. No model found in gs://REDACTED/model/test-folder/model.pb.
```

Did a little bit of debugging and seems like this line 342 is not evaluated to true for single file downloads
```python
if blob.name == bucket_path and len(list(shallow_blobs)) == 0:
```
`list(shallow_blobs)` evaluates to `[<Blob: REDACTED, model/test-folder/model.pb, 1729651649571904>]`

**Comment by spolti** [comment]
Interestingly, this test has passed:
https://github.com/kserve/kserve/pull/4015/files#diff-dcbc48c9c4d2c7da567b988ddf8897c504f1efcd412d7dd4897f9aef609b2cd7R132

Also this image test:
```bash
podman run -it  quay.io/spolti/storage-initializer:4 gs://kfserving-examples/models/xgboost/iris/model.bst /tmp/
WARNING: image platform (linux/amd64) does not match the expected platform (linux/arm64)
2024-10-28 15:39:43.431 1 kserve INFO [initializer-entrypoint:<module>():16] Initializing, args: src_uri [gs://kfserving-examples/models/xgboost/iris/model.bst] dest_path[ [/tmp/]
2024-10-28 15:39:43.432 1 kserve INFO [storage.py:download():64] Copying contents of gs://kfserving-examples/models/xgboost/iris/model.bst to local
2024-10-28 15:39:55.928 1 kserve INFO [storage.py:_download_gcs():349] Downloading single file to: /tmp/model.bst
2024-10-28 15:39:56.141 1 kserve INFO [storage.py:download():110] Successfully copied gs://kfserving-examples/models/xgboost/iris/model.bst to /tmp/
2024-10-28 15:39:56.142 1 kserve INFO [storage.py:download():111] Model downloaded in 12.710291575989686 seconds.

```

can you please try this image with your model?

**Comment by bentohset** [comment]
@spolti Your image works. Seems like it was an oversight on my part - forgot to remove one of my previous edits. Thanks for clarifying.

PR works as intended. LGTM
cc @lizzzcai

**Pull request merged**